mapeditor/scenelayer.cpp

```diff
@@ -55,65 +72,284 @@ void AbstractLayer::redraw()
+	for (QGraphicsItem * existingSector : getAllSectors())
+	{
+		for (auto area : areas)
+		{
+			if (existingSector->sceneBoundingRect().intersects(area))
+			{
+				intersectingSectors.insert(existingSector);
+			}
+		}
+	}
+	redrawSectors(intersectingSectors);
+}
+
+QRectF AbstractViewportLayer::getObjectArea(const CGObjectInstance * object) const
+{
+	auto pos = object->pos;
+	int x = ((pos.x + 1) * tileSize) - (object->getWidth() * tileSize);	//Qt set 0,0 point on the top right corner, CGObjectInstance on the bottom left
+	int y = ((pos.y + 1) * tileSize) - (object->getHeight() * tileSize);
+	QRectF objectArea(x, y, object->getWidth() * tileSize, object->getHeight() * tileSize);
+	return objectArea;
+}
+
+void AbstractViewportLayer::addSector(QGraphicsItem * sector)
+{
+	items->addToGroup(sector);
+}
+
+void AbstractViewportLayer::removeSector(QGraphicsItem * sector)
+{
+	items->removeFromGroup(sector);
+	delete sector;
+}
+
+void AbstractViewportLayer::redrawSectors(std::set<QGraphicsItem *> & sectors)
+{
+	std::set<QGraphicsItem *> sectorsToRemove;
+
+	for (QGraphicsItem * existingSectors : getAllSectors())
+	{
+		for (QGraphicsItem * sector : sectors)
+		{
+			if (existingSectors->sceneBoundingRect().contains(sector->sceneBoundingRect()))
+				sectorsToRemove.insert(existingSectors);
+		}
+	}
+	for (QGraphicsItem * sectorToRemove : sectorsToRemove)
+	{
+		addSector(draw(sectorToRemove->sceneBoundingRect()));
+		removeSector(sectorToRemove);
+	}
+}
+
+const QList<QGraphicsItem *> AbstractViewportLayer::getAllSectors() const	//returning const is necessary to avoid "range-loop might detach Qt container" problem
+{
+	QList<QGraphicsItem *> emptyList;
+	return items ? items->childItems() : emptyList;
+}
+
+std::set<QGraphicsItem *> AbstractViewportLayer::getContainingSectors(const std::vector<int3> & tiles) const
+{
+	std::set<QGraphicsItem *> result;
+	for (QGraphicsItem * existingSector : getAllSectors()) {
+		for (const int3 tile : tiles)
+		{
+			if (existingSector->sceneBoundingRect().contains(QPointF(tile.x * tileSize, tile.y * tileSize)))
+			{
+				result.insert(existingSector);
+				break;
+			}
+		}
+	}
+	return result;
+}
+
+std::set<QGraphicsItem *> AbstractViewportLayer::getIntersectingSectors(const std::vector<QRectF> & areas) const
+{
+	std::set<QGraphicsItem *> result;
+	for (QGraphicsItem * existingSector : getAllSectors()) {
+		for (QRectF area : areas)
+		{
+			if (existingSector->sceneBoundingRect().intersects(area))
+			{
+				result.insert(existingSector);
+			}
+		}
+	}
+	return result;
+}
+
+EmptyLayer::EmptyLayer(MapSceneBase * s): AbstractFixedLayer(s)
+{
+	isShown = true;
+}
+
+void EmptyLayer::update()
 {
 	if(!map)
 		return;
-	
-	pixmap.reset(new QPixmap(map->width * 32, map->height * 32));
-	pixmap->fill(Qt::transparent);
-	
-	QPainter painter(pixmap.get());
-	for(int j = 0; j < map->height; ++j)
+
+	pixmap = std::make_unique<QPixmap>(map->width * 32, map->height * 32);
```

Comment:
Would also be great to remove these massive pixmap's which cause absurdly high memory usage when loading large maps.

Not insisting, and not for this PR, so up to you